Fix handling of empty types in patterns.

canndrew opened the pull request:
Fix for #12609.

rust-highfive commented:
r? @nikomatsakis

(rust_highfive has picked a reviewer for you, use r? to override)

ranma42 commented on src/librustc_const_eval/_match.rs:
```diff
@@ -532,19 +572,24 @@ pub fn is_useful<'a, 'tcx>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
+                ConstructWitness => UsefulWithWitness(vec![Witness(vec![])]),
+                LeaveOutWitness => Useful,
+            }
+        }
```
Style: usually `else` is on the same line as the closing bracket.

oli-obk commented on src/librustc_const_eval/diagnostics.rs:
```diff
@@ -28,7 +28,7 @@ For example, the following `match` block has too many arms:
 match Some(0) {
     Some(bar) => {/* ... */}
-    None => {/* ... */}
+    x => {/* ... */} // This handles the `None` case
```
Why is it necessary to change this example?

oli-obk replied:
ah nevermind, because this triggers for E0001 and not for `unreachable_patterns`

bluss commented:
cc @brson because this would be a new (insta-stable?) feature

nikomatsakis commented:
Seems like we need some more test -- for example, I don't see any tests covering the private field cases, nor the "don't need to match variants" code, right? (is that because this is a WIP?)

canndrew commented on src/librustc_const_eval/check_match.rs:
```diff
@@ -354,14 +355,19 @@ fn check_arms<'a, 'tcx>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
-                                                           "unreachable pattern");
-                            err.span_label(pat.span, &"this is an unreachable pattern");
-                            // if we had a catchall pattern, hint at that
+                            // if we had a catchall pattern, raise an error.
```
Not sure, it was @eddyb's suggestion. If that becomes a lint then we don't get to use E0001 at all though.

arielb1 replied:
Don't use E0001 at all.

canndrew replied:
@eddyb Opinion?

canndrew commented:
> Seems like we need some more test -- for example, I don't see any tests covering the private field cases, nor the "don't need to match variants" code, right? (is that because this is a WIP?)

Yep, it should all be there now though.

canndrew commented:
It might be worth @arielb1 having a look at this too, seeing as they wrote the match checking code that I've fiddled with.

canndrew commented:
So a quick summary of these changes:

(Most) unreachable patterns now generate warnings instead of errors. I've added a new `unreachable_patterns` lint rather than reuse `unreachable_code` just so people can be slightly more specific about {en,dis}abling errors.

Unreachable patterns that used to sneak past the pattern-match checks no longer produce an `unreachable_code` warning later on.

I've fixed `is_uninhabited` for enums. Before it used to think that the fields of enum variants were private and so, for example, it would treat `Result<!, !>` as being inhabited on the assumption that the user can't see the `!`s inside. I did this by making `FieldDefData::is_uninhabited_recurse` take an `is_enum: bool` flag which overrides the field's visibility flag. Is this the best way to do this? Maybe the visibility of enum variant fields should be correctly set in the first place.

`is_useful` no longer exits early with `Useful` if it gets a matrix with zero rows. This is incorrect behaviour when empty types are involved. It would be possible to add this check back and short-circuit the rest of the algorithm so long as we also check for empty types and return `NotUseful` if we find one. But for now I've kept the algorithm simple.

`all_constructors(Ty)` only returns constructors that are actually possible.  eg. for `Result<u32, !>` it will only return `Ok`, for `&[!]` it will only return `&[]`, for empty types it will not return anything etc.

Removed the special-casing that made `match uninhabited_expr {}` compile before. This should be handled naturally by the pattern-matching algs now.

Correct type information is carried on dummy wildcard patterns. This was necessary to make the above change but I suspect it's also a more general bugfix. It used to be possible for the pattern-match algs to be handling a dummy `TyError` when they should really know the actual type, even for non-empty matrices.

`is_useful` produces fewer, more general witnesses using wildcards when it can.  eg. instead of returning `(Ok(_), Some(_))` and `(Err(_), Some(_))` it will just return `(_, Some(_))`.

`PatternKind::Variant` now carries `Substs` info.

`rustc_mir::build::Builder::simplify_match_pair` can now handle enums that have multiple variants but only one statically-possible variant. This allows code like `let x: Result<u32, !> = ...; let Ok(y) = x;` to compile by downcasting `x` to its inner `u32`.

bors commented:
:umbrella: The latest upstream changes (presumably #38053) made this pull request unmergeable. Please resolve the merge conflicts.

canndrew commented:
@nikomatsakis this is ready now btw.

@brson any chance I could get a crater run on this?

nikomatsakis commented:
@canndrew sorry for delay. Mozilla all hands didn't leave much time for reviewing. I'm going to be on vacation next week but will try to find time, else perhaps someone else can/should review.

canndrew commented:
@nikomatsakis That's alright. I was just a bit angsty that this would get ignored for a month or so and eventually bit-rot like the last PR for this did. I understand you guys have a very full plate though.

Probably @arielb1 would be the best person to review this?

arielb1 commented on src/test/compile-fail/issue-12116.rs:
```diff
@@ -20,7 +20,7 @@ fn tail(source_list: &IntList) -> IntList {
         &IntList::Cons(val, box ref next_list) => tail(next_list),
         &IntList::Cons(val, box Nil)           => IntList::Cons(val, box Nil),
-//~^ ERROR unreachable pattern
+//~^ ERROR cannot move out of borrowed content
```
This should be `#[deny(unreachable_patterns)]`

arielb1 commented on src/librustc_const_eval/_match.rs:
```diff
@@ -406,23 +389,57 @@ fn missing_constructors(cx: &mut MatchCheckCtxt,
+        ty::TyAdt(def, substs) if def.is_enum() && def.variants.len() != 1 => {
+            def.variants.iter().filter_map(|v| {
+                let mut visited = FxHashSet::default();
+                if v.is_uninhabited_recurse(&mut visited,
```
I'm worried that the recursion here might cause things to be slow. Could you try adding a cache somewhere?

canndrew commented:
I've refactored `is_uninhabited_recurse` to use a cache. It's now called `uninhabited_from` and it calculates the full forest of nodes that a type is visibly uninhabited from and caches that. Eg. if we have this:

    mod mod_a {
        pub struct SecretlyUninhabited {
            _priv: !,
        }
    }
    mod mod_b {
        pub mod mod_c {
            pub struct SecretlyUninhabited {
                _priv: !,
            }
            mod mod_d {
                ...
            }
        }
    }
    struct MyType {
        foo: mod_a::SecretlyUninhabited,
        bar: mod_b::SecretlyUninhabited,
    }

Then `uninhabited_from(MyType)` will give the set of containing `mod_a`, `mod_c` and all their descendants (represented as `[mod_a, mod_c]`). We can then check whether that set is non-empty to see whether a type is uninhabited or we can check whether the set contains a given node to see whether the type if visibly uninhabited from that node.

canndrew commented:
Anyone know what happened to travis there?

eddyb commented:
@canndrew I think Travis has been busted for almost a week now.

canndrew commented:
Is there anything more I can do to help get this reviewed?

nikomatsakis commented:
@canndrew back now, sorry about that.

nikomatsakis commented:
Will review.

nikomatsakis commented:
Sorry for delay. I am doing a bit of reading up on the general algorithm since I'm feeling ill-equipped to judge this patch. =)

canndrew commented:
That's cool.

> `is_useful` no longer exits early with `Useful` if it gets a matrix with zero rows. This is incorrect behaviour when empty types are involved. It would be possible to add this check back and short-circuit the rest of the algorithm so long as we also check for empty types and return `NotUseful` if we find one. But for now I've kept the algorithm simple.
>
> `all_constructors(Ty)` only returns constructors that are actually possible. eg. for `Result<u32, !>` it will only return `Ok`, for `&[!]` it will only return `&[]`, for empty types it will not return anything etc.

Algorithm-wise these should be the only relevant changes and should be enough to make it correctly handle empty types.

nikomatsakis commented on src/test/compile-fail/unreachable-arm.rs:
```diff
@@ -12,7 +12,16 @@
+    match Foo::B(1) {
+        Foo::B(_) | Foo::A(box _, 1) => { }
+        Foo::A(_, 1) => { }
+        _ => { }
```
Why is this test written with "error-pattern" and not `//~ ERROR`? Can we convert it?

canndrew replied:
Done.

nikomatsakis commented on src/test/compile-fail/match-vec-unreachable.rs:
```diff
@@ -9,13 +9,14 @@
     match *x {
-        [a, (2, 3), _] => (),
-        [(1, 2), (2, 3), b] => (), //~ ERROR unreachable pattern
+        [_, (2, 3), _] => (),
```
Why change `a` to `_` here?

canndrew replied:
I was to avoid the `unused_variables` warning. I've allowed it instead.

nikomatsakis replied:
> It was to avoid the unused_variables warning. I've allowed it instead.

OK, it's fine either way, I was just checking.

nikomatsakis commented on src/librustc/ty/inhabitedness.rs:
```diff
@@ -0,0 +1,264 @@
+}
+
+impl<'a, 'gcx, 'tcx> AdtDef {
+    /// Calculate the forrest of DefIds from which this adt is visibly uninhabited.
```
Nit: s/forrest/forest/

nikomatsakis commented on src/librustc/ty/inhabitedness.rs:
```diff
@@ -0,0 +1,264 @@
+/// Represents a set of DefIds closed under the ancestor relation. That is, if
+/// a DefId is in this set then so are all its descendants.
+#[derive(Clone)]
+pub struct DefIdForrest {
```
Nit: Forest?

nikomatsakis commented on src/librustc/ty/inhabitedness.rs:
```diff
@@ -0,0 +1,264 @@
+use ty::{AdtKind, Visibility, DefIdTree};
+use ty::TypeVariants::*;
+
+/// Represents a set of DefIds closed under the ancestor relation. That is, if
```
What "ancestor" relation? These comments are quite clear, but they don't say much about how this is used. From reading the code, this is my hypothesis:

- These def-ids represent a set of modules. 
- When we have some ADT type like `Foo<T, U, !>`, where some fields/variants have type !, we will compute a set of modules that can see that field and hence can deduce that this ADT is uninhabited, and store those in a forest. 

Seem right? If so, I think it'd be great to talk a bit about how this type is used here.

I'd probably also make this inhabitedness/mod.rs and put the forest code into forest.rs. Small modules ftw! :)

nikomatsakis replied:
Ah, I see the ancestor thing is based on the `DefPath` using `DefIdTree` (which -- preexisting, though -- prob ought to have some comments too...)

nikomatsakis commented on src/librustc/ty/inhabitedness.rs:
```diff
@@ -0,0 +1,264 @@
+                is_enum: bool) -> DefIdForrest
+    {
+        let mut data_uninhabitedness = move || self.ty(tcx, substs).uninhabited_from(visited, tcx);
+        if is_enum {
```
This deserves a comment. Why do we care if this is an enum? Because field privacy is inherited for enums? I think we give a suitable `self.vis` nonetheless (that is, if the enum is public, then `self.vis` will be `Public`) -- I'd prefer to remove this `if`, which seems good esp. if we allow `pub(restricted)` on specific variants or fields of variants someday.

canndrew replied:
> I think we give a suitable `self.vis` nonetheless

Nah we don't. So types like `Result<!, !>` end up not being publicly uninhabited. I don't know what's involved in fixing that or how much work it would be so should I leave a FIXME here for now and fix it in a follow-up PR?

nikomatsakis commented on src/librustc/ty/sty.rs:
```diff
@@ -979,29 +979,21 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
-    /// Checks whether a type is uninhabited.
-    /// If `block` is `Some(id)` it also checks that the uninhabited-ness is visible from `id`.
-    pub fn is_uninhabited(&self, block: Option<NodeId>, cx: TyCtxt<'a, 'gcx, 'tcx>) -> bool {
+    /// Checks whether a type is visibly uninhabited from a particular module.
```
Can you expand this comment to give a snippet of Rust code, showing how uninhabitnedness may vary from place to place?

nikomatsakis commented on src/librustc/ty/sty.rs:
```diff
@@ -979,29 +979,21 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
+    /// Note: just because a type is uninhabited, that doesn't mean that it's
+    /// *visibly* uninhabited outside its module. You sometimes may want
+    /// `is_uninhabited_from` instead.
+    pub fn is_uninhabited_anywhere(&self, tcx: TyCtxt<'a, 'gcx, 'tcx>) -> bool {
```
I don't see any callers to this fn -- where is it used? If nowhere, can we just remove it?

canndrew replied:
I can, but I'm pretty sure we *should* be using it. When I implemented `!` I left comments like this scattered throughout the code

    // FIXME(canndrew): This is_never should really be an is_uninhabited

The idea was that in most the places where we checked for the old `FnDiverges` or the new `TyNever` we should really be checking for uninhabited types generally. As far as I'm concerned `!` is just `enum Void` except it has its own syntax and it coerces. So I'm not sure why `!` still gets special treatment in a lot of places throughout the compiler. Example: greping for that comment takes me first to this code in liveness.rs

    hir::ExprCall(ref f, ref args) => {
      // FIXME(canndrew): This is_never should really be an is_uninhabited
      let diverges = !self.ir.tcx.tables().is_method_call(expr.id) &&
          self.ir.tcx.tables().expr_ty_adjusted(&f).fn_ret().0.is_never();
      let succ = if diverges {
          self.s.exit_ln
      } else {
          succ
      };
      let succ = self.propagate_through_exprs(args, succ);
      self.propagate_through_expr(&f, succ)
    }

So it looks like `!` and `Void` get treated differently by liveness even though they effect liveness identically.

---

For now I've removed the function but left a comment hinting at how it could be re-implemented.

nikomatsakis commented:
@canndrew I also pushed a correction to a broken comment I came across

canndrew commented:
Okay, I've fixed the nits and added more comments explaining the `is_uninhabited`/`DefId` stuff.

bors commented:
:umbrella: The latest upstream changes (presumably #38766) made this pull request unmergeable. Please resolve the merge conflicts.

nikomatsakis commented:
r=me after rebase

I apologize again for taking a while to review @canndrew. This is a nice PR.

arielb1 commented:
I don't think this handles associated types properly. Anyone care to come with test scenarios?

nikomatsakis commented:
@arielb1 i.e., you think that the test for whether they are inhabited/uninhabited is wrong?

canndrew commented:
@bors r=nikomatsakis

canndrew commented:
@arielb1 I tested it and it seemed to work, what problem do you think there is?

canndrew commented:
@nikomatsakis This should be ready now.

bors commented:
:pushpin: Commit 70b7bd9 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 70b7bd94cc90c9e7a67a367859c1c2b66d989fda -->

bors commented:
:hourglass: Testing commit 70b7bd9 with merge cdc0d53...

canndrew commented:
@eddyb I figured something like that was going on.

canndrew commented:
Aye aye aye, @bors' cargo tests are failing because E0001 got removed. Need [this](https://github.com/rust-lang/cargo/pull/3497) to be merged.

canndrew commented:
I've added back the E0001 diagnostic (so that the buildbot tests pass) and I've added `## Note: this error code is no longer emitted by the compiler.` to it because that seems to be what we're doing now.

eddyb commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 291c84a has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 291c84aad48d09e31e84045c1ffccf82e7c2d443 -->

bors commented:
:hourglass: Testing commit 291c84a with merge 85c76a0...

eddyb commented:
@bors r=nikomatsakis force

eddyb commented:
@bors retry

eddyb commented:
@bors r=nikomatsakis

bors commented:
:pushpin: Commit 275c19d has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 275c19d5b6a0f5eceb93e60ee314e73909a12faf -->

bors commented:
:hourglass: Testing commit 275c19d with merge 6f1ae66...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.1420), [status-travis](https://travis-ci.org/rust-lang/rust/builds/189369175)
Approved by: nikomatsakis
Pushing 6f1ae663ef21602841173c0b9549347904e3c9a8 to master...

tomaka commented:
I suggest a `relnotes` tag for this PR.

GuillaumeGomez commented:
Done.
